web/models/audit_log.ex

```diff
@@ -0,0 +1,51 @@
+defmodule HexWeb.AuditLog do
+  use HexWeb.Web, :model
+
+  schema "audit_logs" do
+    field :action, :string
+    field :extra_params, :map
+    timestamps(updated_at: false)
+  end
+
+  def create!(conn, action, opts) do
+    build(conn, action, opts)
+    |> HexWeb.Repo.insert!
+  end
+
+  def build(%Plug.Conn{} = conn, action, opts) do
+    extra_params = extract_extra_params(action, conn, opts)
+    %HexWeb.AuditLog{action: action, extra_params: extra_params}
+  end
+
+  defp extract_extra_params("docs.publish", conn, %HexWeb.Release{id: id}) do
+    with_conn(conn, %{release_id: id})
+  end
+  defp extract_extra_params("docs.revert", conn, %HexWeb.Release{id: id}) do
+    with_conn(conn, %{release_id: id})
+  end
+  defp extract_extra_params("key.generate", conn, %HexWeb.Key{name: name}) do
+    with_conn(conn, %{key: name})
+  end
+  defp extract_extra_params("key.remove", conn, %HexWeb.Key{name: name}) do
+    with_conn(conn, %{key: name})
+  end
+  defp extract_extra_params("owner.add", conn, {%HexWeb.Package{id: package_id}, %HexWeb.User{id: user_id}}) do
+    with_conn(conn, %{package_id: package_id, user_id: user_id})
+  end
+  defp extract_extra_params("owner.remove", conn, {%HexWeb.Package{id: package_id}, %HexWeb.User{id: user_id}}) do
+    with_conn(conn, %{package_id: package_id, user_id: user_id})
+  end
+  defp extract_extra_params("release.publish", conn, %HexWeb.Release{id: id}) do
+    with_conn(conn, %{release_id: id})
+  end
+  defp extract_extra_params("release.revert", conn, %HexWeb.Release{id: id}) do
+    with_conn(conn, %{release_id: id})
+  end
+  defp extract_extra_params("user.reset_password", _conn, %HexWeb.User{id: id}) do
+    %{user_id: id}
+  end
```

Comment:
@ericmj yeah, not logging this action would simplify the implementation a bit. However, reset_password was mentioned in the original issue and e.g. github does that too. Again, if we can keep this out of the scope, the implementation should be simpler and schema more consistent